agent/proxycfg/manager.go

```diff
@@ -0,0 +1,273 @@
+			// do that, but we should validate more generally that that is always
+			// true.
+			err := m.ensureProxyServiceLocked(svc, m.state.ServiceToken(svcID))
+			if err != nil {
+				m.logger.Printf("[ERR] failed to watch proxy service %s: %s", svc.ID,
+					err)
+			}
+		}
+
+		// Now see if any proxies were removed
+		for proxyID := range m.proxies {
+			if _, ok := services[proxyID]; !ok {
+				// Remove them
+				m.removeProxyServiceLocked(proxyID)
+			}
+		}
+
+		m.mu.Unlock()
+
+		// Wait for a state change
+		_, ok := <-stateCh
+		if !ok {
+			// Stopped
+			return
+		}
+	}
+}
+
+// ensureProxyServiceLocked adss or changes the proxy to our state.
+func (m *Manager) ensureProxyServiceLocked(ns *structs.NodeService, token string) error {
+	state, ok := m.proxies[ns.ID]
+
+	if ok {
+		if !state.Changed(ns, token) {
+			// No change
+			return nil
+		}
+
+		// We are updating the proxy, close it's old state
+		state.Close()
+	}
+
+	var err error
+	state, err = NewState(m.logger, m.cache, m.source, ns, token)
+	if err != nil {
+		return err
+	}
+	m.proxies[ns.ID] = state
+
+	// Start a goroutine that will wait for changes and broadcast them to watchers.
+	go func() {
+		ch := state.Watch()
+		for {
+			select {
+			case snap, ok := <-ch:
+				if !ok {
+					// State closed
+					return
+				}
+				m.notify(&snap)
+			}
+		}
+	}()
+
+	return nil
+}
+
+// removeProxyService is called when a service deregisteres and frees all
+// resources for that service.
+func (m *Manager) removeProxyServiceLocked(proxyID string) {
+	state, ok := m.proxies[proxyID]
+	if !ok {
+		return
+	}
+
+	// Closing state will let the goroutine we started in Ensure finish since
+	// watch chan is closed.
+	state.Close()
+	delete(m.proxies, proxyID)
+
+	// We intentionally leave potential watchers hanging here - there is no new
+	// config for them and closing their channels might be indistinguishable from
+	// an error that they should retry. We rely for them to eventually give up
+	// (because they are in fact not running any more) and so the watches be
+	// cleaned up naturally.
+}
+
+func (m *Manager) notify(snap *ConfigSnapshot) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	watchers, ok := m.watchers[snap.ProxyID]
+	if !ok {
+		return
+	}
+
+	for _, ch := range watchers {
+		// Attempt delivery but don't let slow consumers block us forever. They
+		// might miss updates but it's better than breaking everything.
+		//
+		// TODO(banks): should we close their chan here to force them to eventually
+		// notice they are too slow? Not sure if it really helps.
+		select {
+		case ch <- snap:
+		case <-time.After(100 * time.Millisecond):
+		}
+	}
+}
+
+// Watch registers a watch on a proxy. It might not exist yet in which case this
+// will not fail, but no updates will be delivered until the proxy is
+// registered. If there is already a valid snapshot in memory, it will be
+// delivered immediately.
+func (m *Manager) Watch(proxyID string) (<-chan *ConfigSnapshot, func()) {
```

Comment:
For non-test code, I tend to prefer returning an `io.Closer` instead of a raw `func()`. I feel it is more idiomatic, shows intent, and gives us flexibility to introduce errors later. I try to only use the `func()` return for tests. You can make a dumb implementation as shown below.

```go
type funcCloser func()
func(f *funcCloser) Close() error { (*f)(); return nil }
```

Reply:
The idiom I was shooting for here is the "cancel func" pattern which std lib uses in a few places, notably `context.WithCancel`.

Returning a `io.Closer` is OK but it feels heavy weight/unnecessary to make the caller reason about what that thing is. I think `io.Closer` makes 100% sense if you are returning an object that is actually useful for something but should be Closed when done (like a file), but I'm not sure what value it adds in this case? I guess we could return a specific object like: `Watcher { C  <-chan *ConfigSnapshot }` that implements io.Closer. Do you prefer that? Mixing "cancel func" and "dummy io.Closer" seems like an odd middle ground to me 😄 .

I'd also be fine with changing this to accept a `context.Context` and allow the caller to signal done by cancelling that, but it is more lines of code at the call site in most cases to create a context with a cancel func (and the same pattern).

I'm also happy to do either of: name the return to document that it's a canceller, or create a `type CancelFunc func()` to make the intent clear.

I'm not super against your proposal, just not sure I see the value or the pattern you suggest used much elsewhere.

Reply:
Thats fair, I forgot that `context` introduced that into the stdlib. Either way then is fine! 

Introducing a context feels okay too but I agree its not necessary.

The only reason I like the closer route is that the types better signal intent whereas a `func()` doesn't. But I forgot that `context` had to come around and change that for me. :) So feel free to keep as is, add a new type (like `CancelFunc`) or whatever you want.